invoices/update.go

```diff
@@ -90,10 +90,19 @@ func updateInvoice(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
 		}
 	}
 
-	if ctx.mpp == nil {
+	// If this is a keysend payment, then only then will we allow the
+	// legacy payment path to be used.
+	if _, ok := ctx.customRecords[record.KeySendType]; ok {
 		return updateLegacy(ctx, inv)
 	}
 
+	// If the MPP payload wasn't specified, and this isn't a keysend
+	// payment, then we'll return an error that the payment addr was wrong,
+	// since we require it for improved e2e payment security.
+	if ctx.mpp == nil {
+		return nil, ctx.failRes(ResultAddressMismatch), nil
```

Comment:
i'm not sure this is the right error to return. in this case the sender sent a legacy payment, but they are getting errors about a feature they don't have implemented (otherwise they would have used mpp). presumably if they haven't implemented mpp they haven't implemented mpp errors, so this will be interpreted as unknown. perhaps incorrect-payment-details is more appropriate? should probably coordinate with the other implementations to ensure we all agree on how to handle it

Reply:
These errors are mainly just for proper internal logging and tests. The only errors we ever return that're soured from the invoice registry are: `lnwire.FailMPPTimeout` and `IncorrectPaymentDetails`: https://github.com/lightningnetwork/lnd/blob/master/htlcswitch/link.go#L1257